Some new features, including auto hide and pressure border for showing dock

dgsasha opened the pull request:
Hopefully I didn't miss any objects or signals that need to be destroyed when the extension is disabled. Other than that possibility, these changes seem to function well, although they should still be tested more.

dgsasha commented on extension.js:
```diff
@@ -330,10 +405,23 @@ class Extension {
+        // refresh again after overview is hidden or else the dock size is messed up
+        // this is only noticeable with auto hide enabled
+        // todo: find a better way to fix this issue
+        this.overview_hidden = Main.overview.connect('hidden', this._dock_refresh.bind(this));
```
Having the dock refresh twice in a row seems unnecessary, and honestly I think the dock might be refreshing too often. This is the only way I could figure out how to fix the issue of the dock being the wrong size after exiting the overview though. If someone wants to replicate the issue, just open and close the overview with autohide enabled and you will notice that the window overlap causes the dock to hide when it is about 12 pixels above the dock, instead of when it actually touches the dock. This is because the dock gets set to the wrong size during the refresh that happens when exiting the overview.

dgsasha commented:
Most of the issues should be fixed now, added support for Desktop Icons NG and fixed the excessive refreshing. Tested on GNOME 40 and everything seemed to work but I couldn't test the pressure barrier due to GNOME boxes not supporting it.

fthx commented:
Please make a list of issues we'll have to solve.
I suggest to make (when I'll have some time) 1 bugfix version then 2 features version BUT carefully checking each change to avoid some cross-origin new bugs, very hard to fix.

dgsasha commented:
The issues I have observed are as follows:
1. The dock position is messed up when launching a newly installed app and always show is enabled. I have fixed this and will be uploading the fix soon.
3. When locking and unlocking the screen, if a window overlaps the dock, the dock will not be hidden after unlocking. I haven't started working on a fix for this.
4. There are some other very rare issues where the dock size or position are messed up. I've noticed this can happen when unlocking my computer, but I can't consistently replicate the issue. I think I have this fixed though.
5. In some cases, like when running in a virtual machine, the pressure barrier won't work. I have fixed this by making the pressure barrier optional so VM users can still use this extension.

dgsasha commented:
All of the issues I mentioned should be fixed now.

dgsasha commented:
This isn't ready. There is still a bug where the dock size can get messed up after opening a newly installed app, and occasionally the dock autohide can become unresponsive and the dock gets stuck on the screen. The first bug can be eliminated by not refreshing the dock when exiting the overview when launching a newly installed app, or by refreshing the dock again after the overview is hidden. I thought I fixed it already but I guess I didn't. The second bug is probably due to some error causing a crash but I'm not really sure.

dgsasha commented:
I've been busy lately and haven't had time to work on this, I will try to work on it this weekend if possible.

fthx commented:
Hi, I still do not have the time to look at this now.
:-(

dgsasha commented:
Update: For some reason I can't reproduce either of the issues I previously mentioned. I'm wondering if I just didn't have the newest version of my fork installed on my laptop, so I was encountering issues I had already fixed. When ever you have time to look at this you might just want to ignore my last comment. I now think that I was previously wrong and those issues don't actually exist anymore.

fthx commented:
I'll have a look to all that soon for GS 45 version.
